### PR TITLE
Switch to opCost with costedValue

### PR DESCRIPTION
**Pull request by greenhat**
- [x] remove `opCost` without costedValue completely (https://github.com/scalan/special/pull/20);

### CODE REVIEW

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 1]
```diff
@@ -648,9 +648,12 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
 
-      case CostedM.cost(Def(CCostedCollCtor(_, costs, _, accCost))) => opCost(Seq(accCost), costs.sum(intPlusMonoid))
-      case CostedM.cost(Def(CCostedOptionCtor(_, costOpt, _, accCost))) => opCost(Seq(accCost), costOpt.getOrElse(Thunk(0)))
-      case CostedM.cost(Def(CCostedPairCtor(l, r, accCost))) => opCost(Seq(accCost), l.cost + r.cost)
+      case CostedM.cost(Def(CCostedCollCtor(values, costs, _, accCost))) =>
+        opCost(values, Seq(accCost), costs.sum(intPlusMonoid))
+      case CostedM.cost(Def(CCostedOptionCtor(v, costOpt, _, accCost))) =>
+        opCost(v, Seq(accCost), costOpt.getOrElse(Thunk(0)))
+      case CostedM.cost(Def(CCostedPairCtor(l, r, accCost))) =>
+        opCost(l.value, Seq(accCost), l.cost + r.cost)
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: @greenhat for the `CostedPairCtor` case try this `opCost(Pair(l.value, r.value), Seq(accCost), l.cost + r.cost)`
This Pair is more accurate value which is costed

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 2]
```diff
@@ -1260,7 +1263,7 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
       case CreateProveDlog(In(_v)) =>
         val vC = asRep[Costed[GroupElement]](_v)
         val resV: Rep[SigmaProp] = sigmaDslBuilder.proveDlog(vC.value)
-        val cost = opCost(Seq(vC.cost), costOfDHTuple)
+        val cost = opCost(vC.value, Seq(vC.cost), costOfDHTuple)
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: `resV` should be used

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 3]
```diff
@@ -1269,21 +1272,21 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
-        val cost = opCost(Seq(gvC.cost, hvC.cost, uvC.cost, vvC.cost), costOfDHTuple)
+        val cost = opCost(resV, Seq(gvC.cost, hvC.cost, uvC.cost, vvC.cost), costOfDHTuple)
         RCCostedPrim(resV, cost, mkSizeSigmaProp(gvC.size.dataSize * 4L))
 
       case sigmastate.Exponentiate(In(_l), In(_r)) =>
         val l = asRep[Costed[GroupElement]](_l)
         val r = asRep[Costed[BigInt]](_r)
         val value = l.value.exp(r.value)
-        val cost = opCost(Seq(l.cost, r.cost), costOf(node))
+        val cost = opCost(value, Seq(l.cost, r.cost), costOf(node))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: This is correct

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 4]
```diff
@@ -1269,21 +1272,21 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
         RCCostedPrim(value, cost, SizeGroupElement)
 
       case sigmastate.MultiplyGroup(In(_l), In(_r)) =>
         val l = asRep[Costed[GroupElement]](_l)
         val r = asRep[Costed[GroupElement]](_r)
         val value = l.value.multiply(r.value)
-        val cost = opCost(Seq(l.cost, r.cost), costOf(node))
+        val cost = opCost(value, Seq(l.cost, r.cost), costOf(node))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: This is also correct, so what goes to RCCostedXXX, should also be used in opCost

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 5]
```diff
@@ -1338,12 +1341,13 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
         }
 
       case Values.Tuple(InSeq(Seq(x, y))) =>
-        RCCostedPair(x, y, opCost(Seq(x.cost, y.cost), CostTable.newPairValueCost))
+        RCCostedPair(x, y, opCost(x.value, Seq(x.cost, y.cost), CostTable.newPairValueCost))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: The same here, try `Pair(x.value, y.value)` in opCost first argument

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 6]
```diff
@@ -1489,26 +1493,26 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
         xs.elem.eVal.asInstanceOf[Any] match {
           case ce: CollElem[a,_] =>
             val xsC = asRep[Costed[Coll[a]]](xs)
             val v = xsC.value.length
-            RCCostedPrim(v, opCost(Seq(xsC.cost), costOf(node)), SizeInt)
+            RCCostedPrim(v, opCost(v, Seq(xsC.cost), costOf(node)), SizeInt)
           case se: StructElem[_] =>
             val xsC = asRep[Costed[Struct]](xs)
-            RCCostedPrim(se.fields.length, opCost(Seq(xsC.cost), costOf(node)), SizeInt)
+            RCCostedPrim(se.fields.length, opCost(xsC.value, Seq(xsC.cost), costOf(node)), SizeInt)
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: `se.fields.length` should go to opCost, and remember this is graph building, so compiler cannot do common subexpression optimization, so better introduce local val and use it twice.

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 7]
```diff
@@ -1489,26 +1493,26 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
           case pe: PairElem[a,b] =>
             val xsC = asRep[Costed[(a,b)]](xs)
-            RCCostedPrim(2, opCost(Seq(xsC.cost), costOf(node)), SizeInt)
+            RCCostedPrim(2, opCost(xsC.value, Seq(xsC.cost), costOf(node)), SizeInt)
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: Hint
```
val v: Rep[Int] = 2
RCCostedPrim(v, opCost(v, ... ),...)
```

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala [hunk 8]
```diff
@@ -1521,44 +1525,47 @@ trait RuntimeCosting extends CostingRules with DataCosting with Slicing { IR: Ev
...
         val sBox = tryCast[SizeBox](boxC.size)
         val bytes = boxC.value.propositionBytes
-        mkCostedColl(bytes, sBox.propositionBytes.dataSize.toInt, opCost(Seq(boxC.cost), costOf(node)))
+        mkCostedColl(bytes, sBox.propositionBytes.dataSize.toInt, opCost(bytes, Seq(boxC.cost), costOf(node)))
       case utxo.ExtractBytes(In(box)) =>
         val boxC = asRep[Costed[Box]](box)
         val sBox = tryCast[SizeBox](boxC.size)
         val bytes = boxC.value.bytes
-        mkCostedColl(bytes, sBox.bytes.dataSize.toInt, opCost(Seq(boxC.cost), costOf(node)))
+        mkCostedColl(bytes, sBox.bytes.dataSize.toInt, opCost(bytes, Seq(boxC.cost), costOf(node)))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/RuntimeCosting.scala]: this is ok, because bytes go directly ito RCCostedColl inside mkCostedColl

**Comment by aslesarenko** [comment]
@greenhat please also update dependency to the new master of Special

File: src/main/scala/sigmastate/eval/CostingRules.scala [hunk 9]
```diff
@@ -201,13 +201,13 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
     def sizeOfArgs = args.foldLeft(obj.size.dataSize)({ case (s, e) => s + e.size.dataSize })
 
     def constantSizeProperyAccess[R](prop: Rep[T] => Rep[R]): RCosted[R] =
-      withConstantSize(prop(obj.value), opCost(costOfArgs, selectFieldCost))
+      withConstantSize(prop(obj.value), opCost(obj.value, costOfArgs, selectFieldCost))
```
[the remark this answers is not in the document]
Comment (greenhat) [reply src/main/scala/sigmastate/eval/CostingRules.scala]: Thanks. Fixed.

File: src/main/scala/sigmastate/eval/CostingRules.scala [hunk 10]
```diff
@@ -219,25 +219,25 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
...
+      RCCostedOption(v, SOME(itemCost), asSizeOption(s).sizeOpt, opCost(obj.value, costOfArgs, selectFieldCost))
     }
 
     def defaultCollProperyAccess[R: Elem](prop: Rep[T] => RColl[R], propSize: RSize[T] => RSize[Coll[R]], itemCost: Rep[Int]): RCostedColl[R] = {
       val v = prop(obj.value)
       val s = propSize(obj.size)
       val sizes = asSizeColl(s).sizes
       val costs = colBuilder.replicate(sizes.length, itemCost)
-      RCCostedColl(v, costs, sizes, opCost(costOfArgs, selectFieldCost))
+      RCCostedColl(v, costs, sizes, opCost(obj.value, costOfArgs, selectFieldCost))
```
[the remark this answers is not in the document]
Comment (greenhat) [reply src/main/scala/sigmastate/eval/CostingRules.scala]: Fixed.

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala (hunk 1 above)
Comment (greenhat) [reply src/main/scala/sigmastate/eval/RuntimeCosting.scala]: Fixed.

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala (hunk 2 above)
Comment (greenhat) [reply src/main/scala/sigmastate/eval/RuntimeCosting.scala]: Fixed.

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala (hunk 5 above)
Comment (greenhat) [reply src/main/scala/sigmastate/eval/RuntimeCosting.scala]: Fixed.

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala (hunk 6 above)
Comment (greenhat) [reply src/main/scala/sigmastate/eval/RuntimeCosting.scala]: Fixed.

File: src/main/scala/sigmastate/eval/RuntimeCosting.scala (hunk 7 above)
Comment (greenhat) [reply src/main/scala/sigmastate/eval/RuntimeCosting.scala]: Thanks! Fixed.

File: src/main/scala/sigmastate/eval/CostingRules.scala [hunk 11]
```diff
@@ -201,13 +201,13 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
     def sizeOfArgs = args.foldLeft(obj.size.dataSize)({ case (s, e) => s + e.size.dataSize })
 
     def constantSizeProperyAccess[R](prop: Rep[T] => Rep[R]): RCosted[R] =
-      withConstantSize(prop(obj.value), opCost(costOfArgs, selectFieldCost))
+      withConstantSize(prop(obj.value), opCost(obj.value, costOfArgs, selectFieldCost))
 
     def knownSizeProperyAccess[R](prop: Rep[T] => Rep[R], size: RSize[R]): RCosted[R] =
-      RCCostedPrim(prop(obj.value), opCost(costOfArgs, selectFieldCost), size)
+      RCCostedPrim(prop(obj.value), opCost(obj.value, costOfArgs, selectFieldCost), size)
```
[the remark this answers is not in the document]
Comment (greenhat) [reply src/main/scala/sigmastate/eval/CostingRules.scala]: Done. https://github.com/scalan/special/pull/21

File: src/main/scala/sigmastate/eval/CostingRules.scala [hunk 12]
```diff
@@ -197,47 +197,53 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
...
-      RCCostedPrim(prop(obj.value), opCost(costOfArgs, selectFieldCost), size)
+    def knownSizePropertyAccess[R](prop: Rep[T] => Rep[R], size: RSize[R]): RCosted[R] = {
+      val value = prop(obj.value)
+      RCCostedPrim(value, opCost(value, costOfArgs, selectFieldCost), size)
+    }
 
-    def knownLengthCollProperyAccess[R](prop: Rep[T] => Rep[Coll[R]], len: Rep[Int]): Rep[CostedColl[R]] =
-      mkCostedColl(prop(obj.value), len, opCost(costOfArgs, selectFieldCost))
+    def knownLengthCollPropertyAccess[R](prop: Rep[T] => Rep[Coll[R]], len: Rep[Int]): Rep[CostedColl[R]] =
+      mkCostedColl(prop(obj.value), len, opCost(obj.value, costOfArgs, selectFieldCost))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/CostingRules.scala]: @greenhat here you also need to introduce `val value = prop(obj.value)` and use it in opCost.
This means you should also add asserts to RCCostedColl, RCCostedPair, RCCostedOption to catch such cases.
The guiding principle is the following. Every `value` which is wrapped with RCCostedXXX, IF it comes with OpCost node, then this OpCost node should be unique for that specific value. 
This uniqueness is achievied by taking Id of the `value`.

Comment (greenhat) [reply src/main/scala/sigmastate/eval/CostingRules.scala]: Thanks! Done.

File: src/main/scala/sigmastate/eval/CostingRules.scala [hunk 13]
```diff
@@ -197,47 +197,53 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
...
+    def defaultPropertyAccess[R](prop: Rep[T] => Rep[R], propSize: RSize[T] => RSize[R]): RCosted[R] = {
+      val value = prop(obj.value)
+      RCCostedPrim(value, opCost(value, costOfArgs, selectFieldCost), propSize(obj.size))
+    }
 
-    def defaultOptionProperyAccess[R: Elem](prop: Rep[T] => ROption[R], propSize: RSize[T] => RSize[WOption[R]], itemCost: Rep[Int]): RCostedOption[R] = {
+    def defaultOptionPropertyAccess[R: Elem](prop: Rep[T] => ROption[R], propSize: RSize[T] => RSize[WOption[R]], itemCost: Rep[Int]): RCostedOption[R] = {
       val v = prop(obj.value)
       val s = propSize(obj.size)
-      RCCostedOption(v, SOME(itemCost), asSizeOption(s).sizeOpt, opCost(costOfArgs, selectFieldCost))
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/CostingRules.scala]: @greenhat v should be used in opCost. Why it was not caught by asserts?

File: src/main/scala/sigmastate/eval/CostingRules.scala [hunk 14]
```diff
@@ -547,34 +557,38 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
           })
           val values = obj.value.flatMap(f)
           val costs = colBuilder.replicate(sizes.length, 0)
-          RCCostedColl(values, costs, sizes, opCost(costOfArgs, costOf(method)))
+          RCCostedColl(values, costs, sizes, opCost(obj.value, costOfArgs, costOf(method)))
         case _ =>
```
Comment (aslesarenko) [inline src/main/scala/sigmastate/eval/CostingRules.scala]: @greenhat values should be used in opCost

File: src/main/scala/sigmastate/eval/CostingRules.scala (hunk 13 above)
Comment (greenhat) [reply src/main/scala/sigmastate/eval/CostingRules.scala]: Thanks for spotting this one! I failed at `assert` in `special`(checked the wrong cost).

File: src/main/scala/sigmastate/eval/CostingRules.scala (hunk 14 above)
Comment (greenhat) [reply src/main/scala/sigmastate/eval/CostingRules.scala]: Thanks for spotting this one as well! Another incorrect `assert` in `special` is fixed.

**Pull request merged**